File: src/main/java/io/austinzhu/type/bifunctor/Profunctor.java
package io.austinzhu.type.bifunctor;

import io.austinzhu.type.Function;
import io.austinzhu.type.Type;

public interface Profunctor<A extends Type<A>, B extends Type<B>> {

    <A1 extends Type<A1>, B1 extends Type<B1>> Profunctor<A1, B1> dimap(Function<A1, A> f, Function<B, B1> g, Profunctor<A, B> F);

    <A1 extends Type<A1>> Profunctor<A1, B> lmap(Function<A1, A> f, Profunctor<A, B> F);

    <B1 extends Type<B1>> Profunctor<A, B1> rmap(Function<B, B1> g, Profunctor<A, B> F);
}

File: src/main/scala/dev/austinzhu/pattern/bifunctor/Profunctor.scala
package dev.austinzhu.pattern.bifunctor

trait Profunctor[F[_, _]] {

  def dimap[A1, A, B, B1](f: A1 => A, g: B => B1, fab: F[A, B]): F[A1, B1]

  def lmap[A1, A, B](f: A1 => A, fab: F[A, B]): F[A1, B]

  def rmap[A, B, B1](g: B => B1, fab: F[A, B]): F[A, B1]
}